kategory-docs/docs/docs/datatypes/statet/README.md

```diff
@@ -6,4 +6,142 @@ permalink: /docs/datatypes/statet/
+
+In the most basic of scenarios, we'll only be dealing with one monad at a time making our lives nice and easy. However, it's not uncommon to get into scenarios where some function calls will return `Either<Error, A>`, and others will return `State<S, A>`.
+
+So let's rewrite the example of [`State` docs]({{ '/docs/datatypes/state' | relative_url }}), but instead of representing the `Stack` as an optional `NonEmptyList` let's represent it as a `List`.
+
+```kotlin:ank:silent
+import kategory.*
+
+typealias Stack = List<String>
+
+fun pop(stack: Stack): Tuple2<Stack, String> = stack.first().let {
+    stack.drop(1) toT it
+}
+
+fun push(s: String, stack: Stack): Tuple2<Stack, Unit> =
+        listOf(s, *stack.toTypedArray()) toT Unit
+
+fun stackOperations(stack: Stack): Tuple2<Stack, String> {
+    val (s1, _) = push("a", stack)
+    val (s2, _) = pop(s1)
+    return pop(s2)
+}
+```
+
+```kotlin:ank
+stackOperations(listOf("hello", "world", "!"))
+```
+
+But if we now `pop` an empty `Stack` it will result in `java.util.NoSuchElementException: List is empty.`.
+
+```kotlin
+_stackOperations(listOf()) //java.util.NoSuchElementException: List is empty.
+```
+
+Luckily Kategory offers some nice solutions [`Functional Error Handling` docs]({{ '/docs/patterns/error_handling' | relative_url }}).
+Now we can model our error domain with ease.
+
+```kotlin:ank:silent
+sealed class StackError {
+    object StackEmpty : StackError()
+}
+
+typealias StackEmpty = StackError.StackEmpty
+
+fun popE(stack: Stack): Either<StackError, Tuple2<Stack, String>> =
+        if (stack.isEmpty()) StackEmpty.left()
+        else stack.first().let {
+            stack.drop(1) toT it
+        }.right()
+
+fun pushE(s: String, stack: Stack): Either<StackError, Tuple2<Stack, Unit>> =
+        (listOf(s, *stack.toTypedArray()) toT Unit).right()
+
+fun stackOperationsE(stack: Stack): Either<StackError, Tuple2<Stack, String>> {
+    return pushE("a", stack).flatMap { (s1, _) ->
+        popE(s1).flatMap { (s2, _) ->
+            popE(s2)
+        }
+    }
+}
+```
+```kotlin:ank
+stackOperationsE(listOf("hello", "world", "!"))
+```
+```kotlin:ank
+stackOperationsE(listOf())
+```
+
+As is immediately clear, this code while properly modelling the errors, has become more complex but our signature now represents a simple `Stack` as a `List` with an error domain.
+Let's refactor our manual state management in the form of `(S) -> Tuple2<S, A>` to `State`.
+
+So what we want is a return type that represents `Either` a `StackError` or a certain `State` of `Stack.` When working with `State` we don't pass around `Stack` anymore, so there is no parameter to check if the `Stack` is empty.
+
+```kotlin:ank:silent
+fun _popS(): Either<StackError, StateT<IdHK, Stack, String>> = TODO()
```

Comment:
wouldn't be `State` and not `StateT` here ? Seems like a typo

Reply:
they are the same, `type State<S, A> = StateT<IdHK, S, A>` This is the same funny bizness as in the Reader vs ReaderT. It needs to be standardize for simplest usage in the same way in both cases. I still consider this APIs around State, Reader and in general any simplification parametric to the `Id` monad a part that is broken in Kategory at this moment. If you all agree we should create an issue to fix those api's in the cases of `Reader, Writer, State, ..`

Reply:
This concrete part of the docs is stating the need for StateT when you have a stack like `Either<StackError, State<IdHk, Stack, String>>` so I guess it should explicitly say `State` and not `StateT` since both are different things, even if their usage is equal.

About creating the issue, +1. It's good to standardize to simplify the usage IMO

Reply:
Then it should probably say `Either<StackError, State<Stack String>>` isn't it? just for clarity about the stack we have here. Since having a transformer as the inner class does not make a lot of sense semantically speaking to me.

Reply:
This is not possible see first comment on this PR.